lib/execution_engine2/sdk/EE2Runjob.py

```diff
@@ -457,6 +499,166 @@ def _check_job_arguments(self, jobs, has_parent_job=False):
+        # TODO Use and create a method in sdkmr?
+        msg = (
+            f"Couldn't update job record:{job_that_failed_operation} during retry. Aborting:{job_to_abort}"
+            f"Exception:{exception} "
+        )
+        self._safe_cancel(
+            job_id=job_to_abort,
+            terminated_code=TerminatedCode.terminated_by_server_failure,
+        )
+        # TODO Maybe move this log into multiple so not multiple error messages are generated
+        self.logger.error(msg, exc_info=True, stack_info=True)
+        raise RetryFailureException(msg)
+
+    def _validate_retry_presubmit(self, job_id, as_admin):
+        """Validate retry request before attempting to contact scheduler"""
+        # Check to see if you still have permissions to the job and then optionally the parent job id
+        job = self.sdkmr.get_job_with_permission(
+            job_id, JobPermissions.WRITE, as_admin=as_admin
+        )  # type: Job
+        job_input = job.job_input  # type: JobInput
+
+        parent_job = None
+        if job_input.parent_job_id:
+            parent_job = self.sdkmr.get_job_with_permission(
+                job_input.parent_job_id, JobPermissions.WRITE, as_admin=as_admin
+            )
+
+        if job.batch_job:
+            raise CannotRetryJob(
+                "Cannot retry batch job parents. Must retry individual jobs"
+            )
+
+        if not self._retryable(job.status):
+            raise CannotRetryJob(
+                f"Can only retry a cancelled or errored job_id:{job_id} status:{job.status}"
+            )
+
+        return job, job_input, parent_job
+
+    def retry(self, job_id: str, as_admin=False) -> Dict[str, Optional[str]]:
+        """
+        #TODO Add new job requirements/cgroups as an optional param
+        :param job_id: The main job to retry
+        :param as_admin: Run with admin permission
+        :return: The child job id that has been retried
+        """
+
+        job, job_input, parent_job = self._validate_retry_presubmit(
+            job_id=job_id, as_admin=as_admin
+        )
+
+        # Cannot retry a retried job, you must retry the retry_parent
+        if job.retry_parent:
+            return self.retry(str(job.retry_parent), as_admin=as_admin)
+
+        # Get run job params from db, and inject parent job id, then run it
+        run_job_params = self._get_run_job_params_from_existing_job(
+            job, user_id=self.sdkmr.user_id
+        )
+        # Submit job to job scheduler or fail and not count it as a retry attempt
+        run_job_params[_PARENT_RETRY_JOB_ID] = job_id
+        retry_job_id = self.run(params=run_job_params, as_admin=as_admin)
+
+        # Save that the job has been retried, and increment the count. Notify the parent(s)
+        # 1) Notify the retry_parent that it has been retried
+        try:
+            job.modify(inc__retry_count=1, set__retried=True)
+        except Exception as e:
+            self._db_update_failure(
```

Comment:
I'm trying to think through the failure modes here. If this step fails, then the retry job will be aborted (unless that fails too, in which case we're basically screwed), but will still have the retry parent job field filled in. Will that cause any problems?

Reply:
If this step fails, then the retry parent job doesn't get filled in. But no, I don't see how it would cause any issues as no job states are mutated

Reply:
I meant the `retry_parent` field on the new job still points to the retried parent, even though there effectively wasn't a retry.

Reply:
I think the retry was attempted and failed, its just another error condition like a bad resource requirement, or bad parameters. If the record was created in the DB then it's a retry.

Reply:
@ialarmedalien it'd be good for you to weigh in on this one as well

Reply:
see comments here: https://github.com/kbase/execution_engine2/pull/383#discussion_r638336813

Reply:
We agreed to table this for now and discuss in an ADR: https://github.com/kbase/execution_engine2/pull/367/files#diff-370ecf4c171a1ae480f063fbe22e0dbd2c9a418ba7f8cf51dce663887c74196fR76